### PR TITLE
Updated player stats panel

### PR DESCRIPTION
**Pull request by ikostan**
---
name: Default Pull Request Template
about: Suggesting changes to SkyLockAssault
title: ''
labels: ''
assignees: ''
---

## Description

What does this PR do? (e.g., "Fixes player jump physics in level 2" or "Adds
new enemy AI script")

## Related Issue

Closes #ISSUE_NUMBER (if applicable)

## Changes

- [ ] List key changes here (e.g., "Updated Jump.gd to use Godot 4.4's new Tween
  system")
- [ ] Any breaking changes? (e.g., "Deprecated old signal; migrate to new one")

## Testing

- [ ] Ran the game in Godot v4.5 editor—describe what you tested (e.g., "Jump
  works on Win10 with 60 FPS")
- [ ] Any new unit tests added? (Link to test scene if yes)
- [ ] Screenshots/GIFs if UI-related: (Attach below)

## Checklist

- [ ] Code follows Godot style guide (e.g., snake_case for variables)
- [ ] No console errors in editor/output
- [ ] Ready for review!

## Additional Notes

Anything else? (e.g., "Tested on Win10 64-bit; needs Linux validation")

## Summary by Sourcery

Enhance the player stats system with richer fuel and speed UI feedback and rotor control behavior tied to fuel state.

image

New Features:
- Add a speed bar and label to the player stats panel with dynamic color changes and warning behavior.
- Introduce blinking warning states for fuel and speed labels when fuel is low or speed is too low/high.
- Add reusable helpers to start and stop rotor animations and audio, and tie rotor behavior to fuel depletion.

Enhancements:
- Refactor fuel and speed tracking into parameter dictionaries shared between gameplay logic and UI, with centralized style and color handling for progress bars and labels.
- Improve fuel color gradients to smoothly transition from green through yellow to dark red as fuel depletes, and unify label color retrieval and application.

Tests:
- Update existing fuel and difficulty tests to use the new fuel data structure and bar behavior.
- Add tests covering gradual fuel color transitions down to dark red, rotor behavior with null SFX, and independent blinking behavior for fuel and speed labels.

## Summary by Sourcery

Enhance the player stats system with richer fuel and speed UI feedback and rotor behavior tied to fuel state.

New Features:
- Add a speed bar and label to the player stats panel with dynamic color and warning behavior based on current speed.
- Introduce blinking warning states for fuel and speed labels when fuel is low or speed is outside safe thresholds.
- Provide reusable helpers to start and stop rotor animations and audio, and tie rotor shutdown to fuel depletion.

Enhancements:
- Refactor fuel and speed tracking into shared parameter dictionaries with centralized style and color handling for progress bars and labels.
- Improve fuel color gradients to smoothly transition from green through yellow to dark red as fuel depletes, and unify label color retrieval and application.

Tests:
- Update existing fuel and difficulty tests to use the new fuel data structure and bar behavior.
- Add tests for gradual fuel color transitions down to dark red, rotor behavior with null or missing SFX/animation, and independent blinking behavior for fuel and speed labels.

### CODE REVIEW

File: scripts/player.gd [hunk 1]
```diff
@@ -128,56 +228,128 @@ func _input(event: InputEvent) -> void:
...
+		fuel["bar style"].bg_color = Color.RED.lerp(Color(0.5, 0, 0), fuel_section_factor)  # Example to dark red
 	else:
-		fill_style.bg_color = Color.RED  # Full red
+		fuel["bar style"].bg_color = Color.RED.lerp(Color(0.5, 0, 0), fuel_section_factor)  # Example to dark red
+
+
+func update_speed_bar() -> void:
+	speed["bar"].value = speed["speed"]
+	speed["factor"] = speed["speed"] * 100 / MAX_SPEED
+	speed["bar style"].bg_color = Color.CORAL.lerp(Color.DARK_ORANGE, speed["factor"])  # Example to dark red
```
[the remark this answers is not in the document]
Comment (ikostan) [reply scripts/player.gd]: FIXED:

```speed["factor"] = speed["speed"] / MAX_SPEED  # Speed lerp factor is a 0–1 range```

**Comment by ikostan** [comment]
Working on:

- The blinking label implementation shares global label_color and label_warning_color across both fuel and speed, so when both indicators blink at once they will conflict; store the base and warning colors per label (e.g., in each dictionary) instead of using shared globals.

**Comment by ikostan** [comment]
I won't address following in this PR:

- The new fuel/speed dictionaries rely on magic string keys (including keys with spaces like "bar style") and type mixing, which makes the logic fragile and harder to maintain—consider replacing these with dedicated helper objects/structs or at least constants/enums for keys and consistent typed fields.

I opened a new issue in order to fix it later on, see hetre: [FEATURE] Refactor Fuel and Speed Dictionaries to Dedicated StatManager Class in player.gd #276

**Comment by ikostan** [comment]
@sourcery-ai review

**Comment by ikostan** [comment]
Addressing following issue:

issue (bug_risk): Theme override fallback likely never uses the actual override color.

In the fallback branch you still call label.get_theme_color("font_color", "Label"), which reads from the theme, not the override, even when has_theme_color_override("font_color") is true. That means the override is never actually used and you can stay stuck with the invalid/transparent color, breaking _toggle_label comparisons and warning color behavior. You likely want label.get_theme_color("font_color") (no class) or label.get_theme_color_override("font_color"), depending on the Godot version.

**Comment by ikostan** [comment]
> * The `fuel` and `speed` dictionaries with string keys (including keys like "bar style") make the stat model brittle and hard to refactor; consider wrapping these into a small typed helper class/struct or at least constants for the keys to improve safety and readability.

Won't fix
